### PR TITLE
Add Prefixed database Access [ECR-4159]:

### PR DESCRIPTION
**Pull request by dmitry-timofeev**
## Overview
Add the Prefixed access, that can be created either
from a handle, or from the base access and the namespace.

Clarify IndexAddress documentation and StorageIndex spec,
to allow index pooling across Accesses.

---
See: https://jira.bf.local/browse/ECR-4159

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has Javadoc
- [x] Method preconditions are checked and documented in the Javadoc of the method
- [x] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

### CODE REVIEW

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/database/Prefixed.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    OpenIndexRegistry registry = baseAccess.getOpenIndexes();
+    long handle = nativeCreate(namespace, baseAccess.getAccessNativeHandle());
+    return fromHandleInternal(handle, baseAccess.canModify(), cleaner, registry);
+  }
+
+  /**
+   * Creates a Prefixed native peer given the base access native handle.
+   * @throws RuntimeException if the base access is unsupported; or if the namespace is not valid
+   */
+  private static native long nativeCreate(String namespace, long accessNativeHandle);
```
Comment (vitvakatu) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/database/Prefixed.java]: nit: we usually use (handle, arguments....) order in native methods

Comment (dmitry-timofeev) [reply exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/database/Prefixed.java]: In `create*` a handle is just an argument, it is not a handle to "this".

Comment (vitvakatu) [reply exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/database/Prefixed.java]: Okay, I guess it's the matter of taste

File: exonum-java-binding/core/src/test/java/com/exonum/binding/core/storage/database/PrefixedIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+      // Check both the access and the index are inaccessible
+      assertThrows(IllegalStateException.class, prefixed::getAccessNativeHandle);
+      assertThrows(IllegalStateException.class, index::get);
+    }
+
+    @AfterEach
+    void dropDatabase() throws CloseFailuresException {
+      try {
+        cleaner.close();
```
Comment (bullet-tooth) [inline exonum-java-binding/core/src/test/java/com/exonum/binding/core/storage/database/PrefixedIntegrationTest.java]: NPE if an exception occurs in creating db 🙃

Comment (dmitry-timofeev) [reply exonum-java-binding/core/src/test/java/com/exonum/binding/core/storage/database/PrefixedIntegrationTest.java]: d5daac96f772cb8ac5c128d91cf757f31a159d58

**Pull request closed without merging**